src/Interpreters/InterpreterInsertQuery.cpp

```diff
@@ -414,17 +414,12 @@ BlockIO InterpreterInsertQuery::execute()
             return std::make_shared<MaterializingTransform>(in_header);
         });
 
-        size_t num_select_threads = pipeline.getNumThreads();
         size_t num_insert_threads = std::max_element(out_chains.begin(), out_chains.end(), [&](const auto &a, const auto &b)
         {
             return a.getNumThreads() < b.getNumThreads();
         })->getNumThreads();
         pipeline.addChains(std::move(out_chains));
-
         pipeline.setMaxThreads(num_insert_threads);
-        /// Don't use more threads for insert then for select to reduce memory consumption.
```

Comment:
This is the only change that affects the behaviour in some way. I've removed this check assuming  `settings.parallel_view_processing` was always true and the if was not taken since internally it was always doing parallel view processing.

Reply:
Looking at the failed test it seems it's better to keep the other approach and reduce memory consumption.

Reply:
@KochetovNicolai What do you think about this check? I see that you added it when reworking how the pipelines work and I'm not sure if it would be more appropriate to always have it or to never have it, but I'm more inclined to keep the smaller threads / reduced memory consumption by default and maybe revisit it in the future.

Reply:
We build the same pipeline, the setting `parallel_view_processing` only changes the threads number.
This exact line changes the number of threads. It just tells not to use more than `num_select_threads` if `parallel_view_processing` is disabled. Usually, `num_select_threads = 1` (the exception is when `max_insert_threads > 1`, so that we create dedicated select for each parallel insert). So that, when `parallel_view_processing`, we use `max(chain.getNumThreads())` which is a recommended number of threads for each view.  And most likely single thread otherwise.

Also I expected [this](https://github.com/ClickHouse/ClickHouse/blob/ae5570e6d603717dc8f70a1535c67b4fb54247b5/tests/queries/0_stateless/01275_parallel_mv.sql#L16) to fail.

Reply:
Ok, then I see that `parallel_view_processing=0` is working but just in some cases depending on the insert type and that was throwing me off:

```
CREATE TABLE a (a Int64) ENGINE=Memory;
CREATE TABLE b (a Int64) ENGINE=Memory;
CREATE MATERIALIZED VIEW mv1 TO b AS Select sleep(1) as a FROM a;
CREATE MATERIALIZED VIEW mv2 TO b AS Select sleep(1) as a FROM a;
CREATE MATERIALIZED VIEW mv3 TO b AS Select sleep(1) as a FROM a;
CREATE MATERIALIZED VIEW mv4 TO b AS Select sleep(1) as a FROM a;
CREATE MATERIALIZED VIEW mv5 TO b AS Select sleep(1) as a FROM a;
CREATE MATERIALIZED VIEW mv6 TO b AS Select sleep(1) as a FROM a;
CREATE MATERIALIZED VIEW mv7 TO b AS Select sleep(1) as a FROM a;
CREATE MATERIALIZED VIEW mv8 TO b AS Select sleep(1) as a FROM a;
CREATE MATERIALIZED VIEW mv9 TO b AS Select sleep(1) as a FROM a;
CREATE MATERIALIZED VIEW mv10 TO b AS Select sleep(1) as a FROM a;

```
INSERT INTO a VALUES(1);
```
Takes 1 second to insert.

But:

```
INSERT INTO a SELECT * FROM system.one;
```
Takes 10 seconds.

```
INSERT INTO a SELECT * FROM system.numbers_mt LIMIT 1;
```
Takes 10 seconds too.

But:
```
SET max_insert_threads=10, parallel_view_processing=1;
INSERT INTO a SELECT * FROM system.numbers_mt LIMIT 1;
```
Takes 1 second.


Since I was doing my tests with `VALUES` I thought that this wasn't working at all. Is that behaviour then a bug?

Reply:
Looks like a bug indeed. Technically, `insert` and `insert select` are processed a little bit differently. I suppose `insert` ignores `parallel_view_processing`, `insert select` is not.

Reply:
Ok, let's close this PR since it was obviously wrong and I'll create a ticket about that specific issue. I'll try to have a look at it during the week.

Thanks a lot!